File: leetcode/201-300/220-contains-duplicate-iii/bench/contains_duplicate_iii.py
"""Benchmark workload for LeetCode #220 — Contains Duplicate III (Python; scale lane)."""


def bucket_of(x, w):
    # True floor division. Python's `//` already floors for either sign, which is
    # exactly what the compiled versions' truncating-`/` shift formula computes
    # (verified equal across signs) — so this stays byte-identical to them.
    return x // w


def near_value(m, b, x, t):
    v = m.get(b)
    if v is not None and abs(x - v) <= t:
        return True
    return False


def contains(nums, n, k, t):
    if k <= 0:
        return False
    w = t + 1
    buckets = {}
    for i in range(n):
        x = nums[i]
        b = bucket_of(x, w)
        if near_value(buckets, b, x, t):
            return True
        if near_value(buckets, b - 1, x, t):
            return True
        if near_value(buckets, b + 1, x, t):
            return True
        buckets[b] = x
        if i >= k:
            old = nums[i - k]
            buckets.pop(bucket_of(old, w), None)
    return False


def main():
    n = 20000
    pairs = 800
    valrange = 8000000
    half = 4000000

    nums = []
    state = 12345
    for _ in range(n):
        state = (state * 1103515245 + 12345) & 2147483647
        nums.append((state % valrange) - half)

    sink = 0
    for p in range(pairs):
        k = 32 + (p % 512)
        t = p % 3
        if contains(nums, n, k, t):
            sink += 1
    print(sink)


main()

File: leetcode/201-300/220-contains-duplicate-iii/bench/contains_duplicate_iii.rs
use std::collections::HashMap;

fn abs_i64(x: i64) -> i64 {
    if x < 0 { -x } else { x }
}

fn bucket_of(x: i64, w: i64) -> i64 {
    if x >= 0 { x / w } else { (x + 1) / w - 1 }
}

fn near_value(m: &HashMap<i64, i64>, b: i64, x: i64, t: i64) -> bool {
    if let Some(&v) = m.get(&b) {
        if abs_i64(x - v) <= t {
            return true;
        }
    }
    false
}

fn contains(nums: &[i64], n: i64, k: i64, t: i64) -> bool {
    if k <= 0 {
        return false;
    }
    let w = t + 1;
    let mut buckets: HashMap<i64, i64> = HashMap::new();
    let mut i = 0i64;
    while i < n {
        let x = nums[i as usize];
        let b = bucket_of(x, w);
        if near_value(&buckets, b, x, t) {
            return true;
        }
        if near_value(&buckets, b - 1, x, t) {
            return true;
        }
        if near_value(&buckets, b + 1, x, t) {
            return true;
        }
        buckets.insert(b, x);
        if i >= k {
            let old = nums[(i - k) as usize];
            buckets.remove(&bucket_of(old, w));
        }
        i += 1;
    }
    false
}

fn main() {
    let n: i64 = 20000;
    let pairs: i64 = 800;
    let valrange: i64 = 8000000;
    let half: i64 = 4000000;

    let mut nums: Vec<i64> = Vec::new();
    let mut state: i64 = 12345;
    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        nums.push((state % valrange) - half);
    }

    let mut sink = 0i64;
    for p in 0..pairs {
        let k = 32 + (p % 512);
        let t = p % 3;
        if contains(&nums, n, k, t) {
            sink += 1;
        }
    }
    println!("{}", sink);
}
